Disable Unix workarounds for Connect'ing to multiple endpoints

stephentoub opened the pull request:
On unix, once connect fails on a socket, that socket becomes unusable for further operations, including additional connect attempts.  This is at direct odds with Socket's instance Connect methods, some of which allow for multiple connect attempts, either due to multiple IPAddresses being provided or due to a string hostname / DnsEndPoint being provided that could then result in multiple IPAddresses to be tried.

We've explored multiple workarounds for this, all of which have problems.  The workaround still in the code involves creating a temporary socket for each address, connecting to it, and if that's successful then immediately disconnecting and connecting with the actual socket.  But that causes mayhem for a server not expecting that pattern, e.g. that fails if the client disconnects and attempts a reconnect, e.g. #5436.

This leaves us with a few choices, none of which are great:
1. Remove the offending Connect instance methods.  Ideally they'd be replaced with static methods, which can be implemented with POSIX-compliant behavior.  But these methods are heavily used and work on Windows.
2. Always throw from the instance Connect methods when there's a possibility that multiple addresses will be tried, e.g. from Connect(IPAddress[], ...) but also from Connect(EndPoint) if a DnsEndPoint is specified.  This will break existing code that tries to run on unix, but it's also predictable in that usage of these APIs will always fail, and developers will know quickly when using a problematic API and move away from it to supported patterns.
3. Throw from the Connect methods only if multiple addresses are actually supplied, e.g. calling Connect(IPAddress[]) with an array of length 1 would work but an array of length 2 would throw. This will allow a lot more existing code to work, but it's also very unpredictable, e.g. if a string host is provided and gets mapped by DNS to a single IPAddress in the test environment but then multiple IPAddresses in the production environment, everything will work fine in test but then fail in production.

I'm torn between (2) and (3).  This commit implements the more conservative (2), as I expect (3) would actually cause serious problems for deployments, but we can loosen from (2) to (3) if that turns out to be better.  In the meantime, we should also seriously explore adding static methods for v1 for the same functionality (developers can also write such code themselves), and write guidance and helpers folks can use in the meantime, e.g. pseudo-code:
```C#
public static async Task<Socket> Connect(
    AddressFamily addressFamily, SocketType socketType, ProtocolType protocolType,
    IPAddress[] addresses, int port)
{
    Exception lastExc = null;
    foreach (IPAddress address in addresses)
    {
        Socket s = new Socket(addressFamily, socketType, protocolType);
        try
        {
            await s.ConnectAsync(address, port).ConfigureAwait(false);
            return s;
        }
        catch (Exception exc)
        {
            lastExc = exc;
        }
    }

    if (lastExc != null) throw lastExc;
    throw new ArgumentException("No addresses provided", "addresses");
}
```

cc: @ericeil, @pgavlin, @davidsh, @CIPop 
Fixes #5829 
Fixes #5436

bartonjs commented:
FWIW, I like option (2) better than option (3) for the reason you gave: It's easy for (3) to work in local or medium scale, then fail in production/large scale where gDNS/DNSrr and other tricks may cause unanticipated multiple-endpoint behaviors.

bartonjs commented:
Test Innerloop OpenSUSE13.2 Release Build and Test please (file in use error on a build log)

bartonjs commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1080,7 +1020,7 @@ public void Connect(IPAddress[] addresses, int port)
             }
         }
-
+        
```
Nit: whitespace

stephentoub commented:
There is an option (4) I neglected to mention:
- If multiple endpoints are supplied (either directly or via DNS), try the first one, and only throw if that's unsuccessful.  As with (3), that would allow some existing code to work, but would similarly have reliability problems if, for example, two servers returned addresses in different orders.  It might be more reliable than (3), though.

bartonjs commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -2438,6 +2381,16 @@ internal IAsyncResult BeginConnect(string host, int port, AsyncCallback requestC
 
+        private static void ThrowIfNotSupportsMultipleConnectAttempts()
+        {
+#pragma warning disable 162 // unreachable code detected
```
Do we still need this? I thought Roslyn didn't complain here...

bartonjs commented on src/System.Net.Sockets/tests/FunctionalTests/ArgumentValidationTests.cs:
```diff
@@ -698,5 +698,85 @@ public void SendToAsync_NullRemoteEndPoint_Throws_ArgumentNull()
+        {
+            using (TcpClient client = new TcpClient())
+            {
+                Assert.Throws<PlatformNotSupportedException>(() => { client.ConnectAsync("localhost", 12345); });
```
TcpClient doesn't extend Socket.  Couldn't TcpClient instance ConnectAsync use the Socket static Connect that works?

stephentoub replied:
It doesn't extend it, but it exposes the socket  via a read/write property immediately after construction.

stephentoub replied:
I guess I could track whether property was accessed and throw from Connect only if it was...?

bartonjs replied:
:(   This one makes me be more torn between "it needs to be predictable" and "but... this is the API I'd expect people to be calling" (translation: I call it, and clearly everyone writes code like me :smile:).

With these two principles in conflict, I think I still land in "it needs to be predictable".  Running a service is 91% about predictability (5% about telemetry, 3% anticipating failures, 1% luck (that the failure that's bound to happen was one you anticipated)).  These percentages are 100% made up on the spot.

bartonjs replied:
> I guess I could track whether property was accessed and throw from Connect only if it was...?

Sounds complicated, but (to me) like it'll let the scenario succeed.  Remember to set it after the connect succeeds :).

ericeil replied:
What if the property is accessed by a debugger?  Wouldn't want that to change the behavior...

stephentoub replied:
I could use Debugger.NotifyOfCrossThreadDependency...?

stephentoub replied:
Except it's not exposed in the contract :frowning: @terrajobst, any reason we can't add https://msdn.microsoft.com/en-us/library/system.diagnostics.debugger.notifyofcrossthreaddependency(v=vs.110).aspx to the System.Diagnostics.Debug contract?

bartonjs commented:
Modulo the \#pragma and whitespace comments the code LGTM.

My gut feel is that the scenario of using TcpClient with a DNS name is common, so I'd like to see that continue to work.  But if the NCL guys say that some other pattern is more common and would still work, I won't object.  (:moneybag::moneybag:)

stephentoub commented:
(A bunch of tests are failing; not sure why these didn't fail locally for me, but I likely need to disable a bunch on unix.)

ericeil commented:
LGTM

davidsh commented:
>My gut feel is that the scenario of using TcpClient with a DNS name is common, so I'd like to see that continue to work. But if the NCL guys say that some other pattern is more common and would still work, I won't object. (:moneybag::moneybag:)

It is a very common scenario to pass in a DNS name to TcpClient.  Using a single IP address is not the norm.  In fact, we recommend to use DNS names as a best practice. Especially when you consider IPv4 vs. IPv6 connectivity and the increasing prevalence of dual-mode sockets.

stephentoub commented:
> It is a very common scenario to pass in a DNS name to TcpClient

Ok.  I'll get at least the most common use cases working.

pgavlin commented:
Personally, I think that we should also be throwing from the single-address Connect() methods if some other Connect() method has already failed. IMO this will result in an easier-to-diagnose failure than ignoring this case (which would eventually result in the connect() call timing out).

stephentoub commented:
> I think that we should also be throwing from the single-address Connect() methods if some other Connect() method has already failed

We can consider that, but I'd like to do that as a separate thing.

pgavlin commented:
> I'd like to do that as a separate thing.

Okay. Filed #5871 to track.

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -2374,6 +2314,8 @@ internal IAsyncResult BeginConnect(EndPoint remoteEP, AsyncCallback callback, ob
                     throw new NotSupportedException(SR.net_invalidversion);
                 }
 
+                ThrowIfNotSupportsMultipleConnectAttempts();
```
Isn't this covered by the change below in `BeginConnect(string host, int port, ...)`?

stephentoub replied:
Yup, removed.

pgavlin commented:
FWIW, the code in `MultipleAddressConnectAsyncResult` (starts on [line 5863](https://github.com/stephentoub/corefx/blob/sockets_fixes/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L5823)) that handles platforms that do not support multiple connection attempts on the same socket can also be removed.

pgavlin commented:
The instance `ConnectAsync` methods (e.g. [line 4319](https://github.com/stephentoub/corefx/blob/sockets_fixes/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4319)) will also need to throw if a `DnsEndPoint` is provided.

stephentoub commented:
> The instance ConnectAsync methods

Thanks, yeah, already have that fixed locally.

stephentoub commented on src/System.Net.Sockets/src/System.Net.Sockets.csproj:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
I'll fix the whitespace changes VS made on my behalf in this file.

stephentoub commented:
Ok.  I've updated the PR.  In addition to what was here previously, I disabled a bunch of PlatformSpecific tests (in the process I cleaned up some duplication with [Theory]), and I re-implemented TcpClient for Unix so that Connect works in situations where the Client property isn't accessed before calling Connect.

stephentoub commented:
@mmitche, that error is back:
ERROR: [WS-CLEANUP] Cannot delete workspace: remote file operation failed: d:\j\workspace\dotnet_corefx\opensuse13.2_release_bld_prtest at hudson.remoting.Channel@2b4b2224:dci-win-fbld-24: java.nio.file.AccessDeniedException: d:\j\workspace\dotnet_corefx\opensuse13.2_release_bld_prtest\Tools\dotnetcli\bin\coreclr.dll

stephentoub commented:
Thanks for the additional feedback, @bartonjs.  I've pushed an additional commit for the small resulting cleanup.

stephentoub commented:
@mmitche, another workspace cleanup error.

stephentoub commented:
Test Innerloop OSX Debug Build and Test please
Test Innerloop OpenSUSE13.2 Release Build and Test please

bartonjs commented:
LGTM

ericeil commented:
Looks good.  Thanks for fixing this!

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -0,0 +1,424 @@
+
+        // Shadow values for storing data set through TcpClient's properties.
+        // We use a separate bool field to store whether the value has been set.
+        // We don't use nullables, due to one of the properties being a refernece type.
```
Nit: s/refernece/reference/

stephentoub replied:
Will fix.

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -0,0 +1,424 @@
+        // We use a separate bool field to store whether the value has been set.
+        // We don't use nullables, due to one of the properties being a refernece type.
+
+        private int _exclusiveAddressUse;
```
Might it make sense to bundle these properties in a lazily-initialized class? It would make this a bit more pay-for-play.

stephentoub replied:
Sure, I'll do that.

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -0,0 +1,424 @@
+            }
+        }
+
+        private void TransferInitializedValuesToSocket(Socket socket)
```
Super nit-pick: I might instead call this `ApplyInitializedOptionsToSocket`.

stephentoub replied:
Will fix.

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -0,0 +1,424 @@
+            {
+                // If we have a client socket, return whether it's connected.
+                // Otherwise as we don't have a socket, by definition it's not.
+                return _clientSocket != null ? _clientSocket.Connected : false;
```
Nit: could just be `_clientSocket != null && _clientSocket.Connected`

stephentoub replied:
Yup, will fix.

pgavlin commented on src/System.Net.Sockets/src/System/Net/Sockets/TCPClient.Unix.cs:
```diff
@@ -0,0 +1,424 @@
+            // Then store the option on to the client socket.  If the client socket isn't yet set, we still want to set
+            // the property onto a socket so we can get validation performed by the underlying system on the option being
+            // set... so, albeit being a bit expensive, we create a temporary socket we can use for validation purposes.
+            Socket s = _clientSocket ?? CreateSocket();
```
Should we not do this validation before storing the value into the shadow property? As written it looks like getting and setting options will work differently than it would if the underlying socket was not temporary.

stephentoub replied:
Hmm, that's a good point. I'll move the setting of the shadow down until after this.

stephentoub commented:
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please
Test Innerloop Ubuntu Debug Build and Test please
Test Innerloop Ubuntu Release Build and Test please

stephentoub commented:
Test Innerloop OSX Debug Build and Test please

stephentoub commented:
Test Innerloop Ubuntu Debug Build and Test please
Test Innerloop Ubuntu Release Build and Test please

stephentoub commented:
Test Innerloop Ubuntu Debug Build and Test please
 Test Innerloop Ubuntu Release Build and Test please
